source/channel_bootstrap.c

```diff
@@ -1160,41 +1182,41 @@ static void s_on_server_channel_on_setup_completed(struct aws_channel *channel,
     if (channel_data->server_connection_args->use_tls) {
         /* incoming callback will be invoked upon the negotiation completion so don't do it
          * here. */
-        if (s_setup_server_tls(channel_data->server_connection_args, channel)) {
+        if (s_setup_server_tls(channel_data, channel)) {
             err_code = aws_last_error();
             goto error;
         }
     } else {
-        channel_data->server_connection_args->incoming_callback(
-            channel_data->server_connection_args->bootstrap,
-            AWS_OP_SUCCESS,
-            channel,
-            channel_data->server_connection_args->user_data);
+        s_server_incoming_callback(channel_data, AWS_OP_SUCCESS, channel);
     }
     return;
 
 error:
     /* shut down the channel */
-    /* TODO: the user incoming callback is not fired, and the shutdown callback should not be fired */
     aws_channel_shutdown(channel, err_code);
 }
 
 static void s_on_server_channel_on_shutdown(struct aws_channel *channel, int error_code, void *user_data) {
     struct server_channel_data *channel_data = user_data;
-
+    struct server_connection_args *args = channel_data->server_connection_args;
     AWS_LOGF_DEBUG(
         AWS_LS_IO_CHANNEL_BOOTSTRAP,
         "id=%p: channel %p shutdown with error %d.",
-        (void *)channel_data->server_connection_args->bootstrap,
+        (void *)args->bootstrap,
         (void *)channel,
         error_code);
 
-    void *server_shutdown_user_data = channel_data->server_connection_args->user_data;
-    struct aws_server_bootstrap *server_bootstrap = channel_data->server_connection_args->bootstrap;
+    void *server_shutdown_user_data = args->user_data;
+    struct aws_server_bootstrap *server_bootstrap = args->bootstrap;
     struct aws_allocator *allocator = server_bootstrap->allocator;
 
-    channel_data->server_connection_args->shutdown_callback(
-        server_bootstrap, error_code, channel, server_shutdown_user_data);
+    if (!channel_data->incoming_called) {
+        error_code = (error_code) ? error_code : AWS_ERROR_UNKNOWN;
+        s_server_incoming_callback(channel_data, error_code, NULL);
+    } else if (channel_data->can_shutdown) {
+        args->shutdown_callback(server_bootstrap, error_code, channel, server_shutdown_user_data);
```

Comment:
There's no need for this `can_shutdown` variable, this can just be an `else {` branch.

If the channel exists, but some error occurs, we **don't want to allow** s_server_incoming_callback to be fired until the channel-shutdown callback.